storage/src/robocorp/storage/_client.py

```diff
@@ -0,0 +1,170 @@
+        return cast(AssetDetails, response.json())
+
+    def delete_asset(self, asset_id: str):
+        """Delete the given asset."""
+        path = quote_plus(asset_id)
+        handler = self._handle_asset_not_found(asset_id)
+        self._client.delete(path, _handle_error=handler)
+
+    def upload_asset(
+        self,
+        asset_id: str,
+        content: bytes,
+        content_type: Optional[ContentType] = None,
+        wait=True,
+        poll_interval=0.5,
+    ):
+        """Upload an asset payload, and optionally wait for it to finish."""
+        if _estimate_base64_size(len(content)) < DATA_LIMIT:
+            LOGGER.info("Content size under data limit, uploading directly")
+            data_content = base64.b64encode(content).decode("ascii")
+            assert len(data_content) < DATA_LIMIT, "Unexpected data size"
+            response = self._create_upload(
+                asset_id,
+                content_type=content_type,
+                data=data_content,
+            )
+        else:
+            LOGGER.info("Content size over data limit, requesting upload URL")
+            response = self._create_upload(asset_id, content_type=content_type)
+            self._client.put(response["upload_url"], data=content, headers={})
+
+        # NOTE (2023-07-04):
+        # The asset upload should be done if the response says so,
+        # but in reality it sometimes takes a while for the value to propagate
+
+        # if response["status"] == "completed":
+        #    return
+
+        if not wait:
+            return
+
+        LOGGER.info("Waiting for asset upload to complete")
+        while True:
+            state = self._get_upload(asset_id, response["id"])
+            status = state["status"]
+
+            if status == "pending":
+                time.sleep(poll_interval)
+                continue
+            elif status == "completed":
+                break
+            elif status == "failed":
+                reason = state.get("reason", "unknown")
+                raise AssetUploadFailed(f"Asset upload failed: {reason}")
+            else:
+                raise AssetUploadFailed(f"Asset upload invalid status: {status}")
+
+    def _create_upload(
+        self,
+        asset_id: str,
+        content_type: Optional[ContentType] = None,
+        data: Optional[str] = None,
+    ) -> AssetUploadResponse:
+        """Create an upload for an existing asset.
+
+        For payloads of less than 5MB you can upload the contents directly on
+        upload creation. For larger payloads you can create the upload and use
+        the returned upload URL for delivering the content.
+        """
+        payload = {}
+        if content_type is not None:
+            payload["content_type"] = content_type
+        if data is not None:
+            payload["data"] = data
+
+        path = url_join(quote_plus(asset_id), "upload")
+        handler = self._handle_asset_not_found(asset_id)
```

Comment:
Is this handler correct? 
i.e.: it seems a bit strange having a `_handle_asset_not_found` when creating an asset.

Reply:
Resolving as it seems there's a slight possibility of it becoming invalid between the create time/upload time.

Reply:
Yes, this upload endpoint is for updating the contents of existing assets, and the creation endpoint is separate. The current codepath _does_ always check that it exists before calling this, but I added that handler for every endpoint where 404 means "asset not found", even if unlikely.